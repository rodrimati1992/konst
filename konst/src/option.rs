//! `const` equivalents of `Option` methods.

/// A const equivalent of [`Option::unwrap`]
///
/// # Example
///
/// ```rust
/// use konst::option::unwrap;
///
/// use std::num::NonZeroUsize;
///
/// const TEN: NonZeroUsize = unwrap!(NonZeroUsize::new(10));
///
/// assert_eq!(TEN.get(), 10);
/// ```
#[doc(inline)]
pub use konst_kernel::opt_unwrap as unwrap;

/// A const equivalent of [`Option::unwrap_or`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[u32] = &[
///     option::unwrap_or!(Some(3), 10000),
///     option::unwrap_or!(None, 5),
/// ];
///
/// assert_eq!(ARR, &[3, 5]);
///
/// ```
///
#[doc(inline)]
pub use konst_kernel::opt_unwrap_or as unwrap_or;

/// A const equivalent of [`Option::unwrap_or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[u32] = &[
///     // You can use a closure-like syntax to run code when the Option argument is None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::unwrap_or_else!(Some(3), || loop{}),
///     option::unwrap_or_else!(None, || 5),
///
///     // You can also pass functions
///     option::unwrap_or_else!(Some(8), thirteen),
///     option::unwrap_or_else!(None, thirteen),
/// ];
///
/// assert_eq!(ARR, &[3, 5, 8, 13]);
///
/// const fn thirteen() -> u32 {
///     13
/// }
/// ```
///
#[doc(inline)]
pub use konst_kernel::opt_unwrap_or_else as unwrap_or_else;

/// A const equivalent of [`Option::ok_or`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Result<u32, u32>] = &[
///     option::ok_or!(Some(3), 10000),
///     option::ok_or!(None, 5),
/// ];
///
/// assert_eq!(ARR, &[Ok(3), Err(5)]);
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_ok_or as ok_or;

/// A const equivalent of [`Option::ok_or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Result<u32, u32>] = &[
///     // You can use a closure-like syntax to run code when the Option argument is None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::ok_or_else!(Some(3), || loop{}),
///     option::ok_or_else!(None, || 5),
///
///     // You can also pass functions
///     option::ok_or_else!(Some(8), thirteen),
///     option::ok_or_else!(None, thirteen),
/// ];
///
/// assert_eq!(ARR, &[Ok(3), Err(5), Ok(8), Err(13)]);
///
/// const fn thirteen() -> u32 {
///     13
/// }
/// ```
#[doc(inline)]
pub use konst_kernel::opt_ok_or_else as ok_or_else;

/// A const equivalent of [`Option::map`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that maps the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::map!(Some(3), |x| x * 3),
///     option::map!(None::<u32>, |_| loop{}),
///
///     // You can also pass functions
///     option::map!(Some(8), double),
///     option::map!(None::<u32>, double),
/// ];
///
/// assert_eq!(ARR, &[Some(9), None, Some(16), None]);
///
/// const fn double(x: u32) -> u32 {
///     x * 2
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_map as map;

/// A const equivalent of [`Option::and_then`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that uses the value in the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::and_then!(Some(3), |x| Some(x * 3)),
///     option::and_then!(Some(3), |_| None),
///     option::and_then!(None::<u32>, |_| loop{}),
///
///     // You can also pass functions
///     option::and_then!(Some(23), checked_sub),
///     option::and_then!(Some(9), checked_sub),
///     option::and_then!(None::<u32>, checked_sub),
/// ];
///
/// assert_eq!(ARR, &[Some(9), None, None, Some(13), None, None]);
///
/// const fn checked_sub(x: u32) -> Option<u32> {
/// # /*
///     x.checked_sub(10)
/// # */
/// #   let (ret, overflowed) = x.overflowing_sub(10);
/// #   if overflowed { None } else { Some(ret) }
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_and_then as and_then;

/// A const equivalent of [`Option::or_else`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that runs on None.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::or_else!(Some(3), || loop{}),
///     option::or_else!(None::<u32>, || Some(5)),
///
///     // You can also pass functions
///     option::or_else!(Some(8), thirteen),
///     option::or_else!(None::<u32>, thirteen),
/// ];
///
/// assert_eq!(ARR, &[Some(3), Some(5), Some(8), Some(13)]);
///
/// const fn thirteen() -> Option<u32> {
///     Some(13)
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_or_else as or_else;

/// A const equivalent of [`Option::flatten`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     option::flatten!(Some(Some(8))),
///     option::flatten!(None),
/// ];
///
/// assert_eq!(ARR, &[Some(8), None]);
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_flatten as flatten;

/// A const equivalent of [`Option::filter`]
///
/// # Example
///
/// ```
/// use konst::option;
///
/// const ARR: &[Option<u32>] = &[
///     // You can use a closure-like syntax to pass code that filters the Some variant.
///     // `return` inside the "closure" returns from the function where this macro is called.
///     option::filter!(Some(0), |&x| x == 0),
///     option::filter!(Some(1), |x| *x == 0),
///     option::filter!(None, |_| loop{}),
///
///     // You can also pass functions
///     option::filter!(Some(3), is_odd),
///     option::filter!(Some(4), is_odd),
///     option::filter!(None, is_odd),
/// ];
///
/// assert_eq!(ARR, &[Some(0), None, None, Some(3), None, None]);
///
/// const fn is_odd(x: &u32) -> bool {
///     *x % 2 == 1
/// }
///
/// ```
#[doc(inline)]
pub use konst_kernel::opt_filter as filter;

/// A const equivalent of the [`Option::copied`] method.
///
/// # Example
///
/// ```rust
/// use konst::option;
///
/// const fn get_last(slice: &[u64]) -> Option<u64> {
///     option::copied(slice.last())
/// }
///
/// assert_eq!(get_last(&[]), None);
/// assert_eq!(get_last(&[16]), Some(16));
/// assert_eq!(get_last(&[3, 5, 8, 13]), Some(13));
///
///
/// ```
pub const fn copied<T: Copy>(opt: Option<&T>) -> Option<T> {
    match opt {
        Some(x) => Some(*x),
        None => None,
    }
}

declare_generic_const! {
    /// Usable to do `[None::<T>; LEN]` when `T` is non-`Copy`.
    ///
    /// As of Rust 1.65.0, `[None::<T>; LEN]` is not valid for non-`Copy` types,
    /// but `[CONST; LEN]` does work, like in the example below.
    ///
    /// # Example
    ///
    /// ```rust
    /// use konst::option::NONE;
    ///
    /// use std::mem::{self, MaybeUninit};
    ///
    /// const TEN: [Option<String>; 10] = [NONE::V; 10];
    ///
    /// let ten = [NONE::<String>::V; 10];
    ///
    /// // the `vec` macro only needs `Option<String>` to be clonable, not Copy.
    /// assert_eq!(vec![None::<String>; 10], TEN);
    ///
    /// assert_eq!(vec![None::<String>; 10], ten);
    ///
    for[T]
    pub const NONE[T]: Option<T> = None;
}
