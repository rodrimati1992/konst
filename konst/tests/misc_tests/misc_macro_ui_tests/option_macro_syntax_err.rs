use konst::option;


const fn whatever0_return<T>() -> T {
    unimplemented!()
}
const fn whatever1_return<T>(a0: impl Sized) -> T {
    std::mem::forget(a0);
    unimplemented!()
}


const _: u8 = option::unwrap_or!();
const _: u8 = option::unwrap_or!(None,);

const _: u8 = option::unwrap_or_else!();
const _: u8 = option::unwrap_or_else!(None,);
const _: u8 = option::unwrap_or_else!(None, |_| 0u8);
const _: u8 = option::unwrap_or_else!(None::<u8>, {whatever0_return});

const _: Result<(), u8> = option::ok_or!();
const _: Result<(), u8> = option::ok_or!(None,);

const _: Result<(), u8> = option::ok_or_else!();
const _: Result<(), u8> = option::ok_or_else!(None::<()>,);
const _: Result<(), u8> = option::ok_or_else!(None::<()>, |_| 0u8);
const _: Result<(), u8> = option::ok_or_else!(None::<()>, {whatever0_return::<u8>});

const _: Option<u8> = option::map!();
const _: Option<u8> = option::map!(None::<()>,);
const _: Option<u8> = option::map!(None::<()>, || 0u8);
const _: Option<u8> = option::map!(None::<()>, {whatever1_return});

const _: Option<u8> = option::and_then!();
const _: Option<u8> = option::and_then!(None::<()>,);
const _: Option<u8> = option::and_then!(None::<()>, || None);
const _: Option<u8> = option::and_then!(None::<()>, {whatever0_return});

const _: Option<u8> = option::or_else!();
const _: Option<u8> = option::or_else!(None::<u8>,);
const _: Option<u8> = option::or_else!(None::<u8>, |_| None);
const _: Option<u8> = option::or_else!(None::<u8>, {whatever0_return});

const _: Option<u8> = option::filter!();
const _: Option<u8> = option::filter!(None::<u8>,);
const _: Option<u8> = option::filter!(None::<u8>, || false);
const _: Option<u8> = option::filter!(None::<u8>, {whatever1_return});

const _: () = {
    let _: &mut u8 = option::get_or_insert!();
    let _: &mut u8 = option::get_or_insert!(&mut None,);
};

const _: () = {
    let _: &mut u8 = option::get_or_insert_with!();
    let _: &mut u8 = option::get_or_insert_with!(&mut None,);
    let _: &mut u8 = option::get_or_insert_with!(&mut None, |_| 0u8);
    let _: &mut u8 = option::get_or_insert_with!(&mut None, ||);
    let _: &mut u8 = option::get_or_insert_with!(&mut None, {whatever0_return});
};

const _: () = {
    let _: &mut u8 = option::insert!();
    let _: &mut u8 = option::insert!(&mut None,);
};

const _: Option<(u8, u16)> = option::zip!();
const _: Option<(u8, u16)> = option::zip!(None,);

const _: bool = option::is_some_and!();
const _: bool = option::is_some_and!(None,);
const _: bool = option::is_some_and!(None, || false);
const _: bool = option::is_some_and!(None, {whatever1_return});

const _: bool = option::is_none_or!();
const _: bool = option::is_none_or!(None,);
const _: bool = option::is_none_or!(None, || false);
const _: bool = option::is_none_or!(None, {whatever1_return});


fn main() {}