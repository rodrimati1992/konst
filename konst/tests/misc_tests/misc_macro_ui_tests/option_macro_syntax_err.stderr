error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:13:15
   |
13 | const _: u8 = option::unwrap_or!();
   |               ^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$e:expr`
  --> src/option.rs
   |
   |     ($e:expr, $v:expr $(,)?) => {
   |      ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:14:39
   |
14 | const _: u8 = option::unwrap_or!(None,);
   |                                       ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$v:expr`
  --> src/option.rs
   |
   |     ($e:expr, $v:expr $(,)?) => {
   |               ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:16:15
   |
16 | const _: u8 = option::unwrap_or_else!();
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$e:expr`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |      ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:17:44
   |
17 | const _: u8 = option::unwrap_or_else!(None,);
   |                                            ^ missing tokens in macro arguments
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |               ^^

error: expected the closure to take no arguments
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:18:15
   |
18 | const _: u8 = option::unwrap_or_else!(None, |_| 0u8);
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::unwrap_or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:19:51
   |
19 | const _: u8 = option::unwrap_or_else!(None::<u8>, {whatever0_return});
   |                                                   ^ no rules expected this token in macro call
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |               ^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:21:27
   |
21 | const _: Result<(), u8> = option::ok_or!();
   |                           ^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$e:expr`
  --> src/option.rs
   |
   |     ($e:expr, $v:expr $(,)?) => {
   |      ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:22:47
   |
22 | const _: Result<(), u8> = option::ok_or!(None,);
   |                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$v:expr`
  --> src/option.rs
   |
   |     ($e:expr, $v:expr $(,)?) => {
   |               ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:24:27
   |
24 | const _: Result<(), u8> = option::ok_or_else!();
   |                           ^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$e:expr`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |      ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:25:58
   |
25 | const _: Result<(), u8> = option::ok_or_else!(None::<()>,);
   |                                                          ^ missing tokens in macro arguments
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |               ^^

error: expected the closure to take no arguments
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:26:27
   |
26 | const _: Result<(), u8> = option::ok_or_else!(None::<()>, |_| 0u8);
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::ok_or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:27:59
   |
27 | const _: Result<(), u8> = option::ok_or_else!(None::<()>, {whatever0_return::<u8>});
   |                                                           ^ no rules expected this token in macro call
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($e:expr, || $v:expr $(,)?) => {
   |               ^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:29:23
   |
29 | const _: Option<u8> = option::map!();
   |                       ^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:30:47
   |
30 | const _: Option<u8> = option::map!(None::<()>,);
   |                                               ^ missing tokens in macro arguments
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |                 ^

error: expected the closure to take a pattern as an argument
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:31:23
   |
31 | const _: Option<u8> = option::map!(None::<()>, || 0u8);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::map` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:32:48
   |
32 | const _: Option<u8> = option::map!(None::<()>, {whatever1_return});
   |                                                ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |                 ^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:34:23
   |
34 | const _: Option<u8> = option::and_then!();
   |                       ^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:35:52
   |
35 | const _: Option<u8> = option::and_then!(None::<()>,);
   |                                                    ^ missing tokens in macro arguments
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |                 ^

error: expected the closure to take a pattern as an argument
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:36:23
   |
36 | const _: Option<u8> = option::and_then!(None::<()>, || None);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::and_then` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:37:53
   |
37 | const _: Option<u8> = option::and_then!(None::<()>, {whatever0_return});
   |                                                     ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $mapper:expr $(,)? ) => {
   |                 ^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:39:23
   |
39 | const _: Option<u8> = option::or_else!();
   |                       ^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, || $mapper:expr $(,)? ) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:40:51
   |
40 | const _: Option<u8> = option::or_else!(None::<u8>,);
   |                                                   ^ missing tokens in macro arguments
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($opt:expr, || $mapper:expr $(,)? ) => {
   |                 ^^

error: expected the closure to take no arguments
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:41:23
   |
41 | const _: Option<u8> = option::or_else!(None::<u8>, |_| None);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::or_else` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:42:52
   |
42 | const _: Option<u8> = option::or_else!(None::<u8>, {whatever0_return});
   |                                                    ^ no rules expected this token in macro call
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($opt:expr, || $mapper:expr $(,)? ) => {
   |                 ^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:44:23
   |
44 | const _: Option<u8> = option::filter!();
   |                       ^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$e:expr`
  --> src/option.rs
   |
   |     ($e:expr, |$param:pat_param| $v:expr $(,)?) => {
   |      ^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:45:50
   |
45 | const _: Option<u8> = option::filter!(None::<u8>,);
   |                                                  ^ missing tokens in macro arguments
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($e:expr, |$param:pat_param| $v:expr $(,)?) => {
   |               ^

error: expected the closure to take a pattern as an argument
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:46:23
   |
46 | const _: Option<u8> = option::filter!(None::<u8>, || false);
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::filter` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:47:51
   |
47 | const _: Option<u8> = option::filter!(None::<u8>, {whatever1_return});
   |                                                   ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($e:expr, |$param:pat_param| $v:expr $(,)?) => {
   |               ^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:50:22
   |
50 |     let _: &mut u8 = option::get_or_insert!();
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, $inserted:expr $(,)?) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:51:55
   |
51 |     let _: &mut u8 = option::get_or_insert!(&mut None,);
   |                                                       ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$inserted:expr`
  --> src/option.rs
   |
   |     ($opt:expr, $inserted:expr $(,)?) => {
   |                 ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:55:22
   |
55 |     let _: &mut u8 = option::get_or_insert_with!();
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, || $default:expr $(,)?) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:56:60
   |
56 |     let _: &mut u8 = option::get_or_insert_with!(&mut None,);
   |                                                            ^ missing tokens in macro arguments
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($opt:expr, || $default:expr $(,)?) => {
   |                 ^^

error: expected the closure to take no arguments
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:57:22
   |
57 |     let _: &mut u8 = option::get_or_insert_with!(&mut None, |_| 0u8);
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::get_or_insert_with` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:58:63
   |
58 |     let _: &mut u8 = option::get_or_insert_with!(&mut None, ||);
   |                                                               ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$default:expr`
  --> src/option.rs
   |
   |     ($opt:expr, || $default:expr $(,)?) => {
   |                    ^^^^^^^^^^^^^

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:59:61
   |
59 |     let _: &mut u8 = option::get_or_insert_with!(&mut None, {whatever0_return});
   |                                                             ^ no rules expected this token in macro call
   |
note: while trying to match `||`
  --> src/option.rs
   |
   |     ($opt:expr, || $default:expr $(,)?) => {
   |                 ^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:63:22
   |
63 |     let _: &mut u8 = option::insert!();
   |                      ^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, $inserted:expr $(,)?) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:64:48
   |
64 |     let _: &mut u8 = option::insert!(&mut None,);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$inserted:expr`
  --> src/option.rs
   |
   |     ($opt:expr, $inserted:expr $(,)?) => {
   |                 ^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:67:30
   |
67 | const _: Option<(u8, u16)> = option::zip!();
   |                              ^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$left:expr`
  --> src/option.rs
   |
   |     ($left:expr, $right:expr) => {
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:68:48
   |
68 | const _: Option<(u8, u16)> = option::zip!(None,);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right:expr`
  --> src/option.rs
   |
   |     ($left:expr, $right:expr) => {
   |                  ^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:70:17
   |
70 | const _: bool = option::is_some_and!();
   |                 ^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:71:43
   |
71 | const _: bool = option::is_some_and!(None,);
   |                                           ^ missing tokens in macro arguments
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |                 ^

error: expected the closure to take a pattern as an argument
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:72:17
   |
72 | const _: bool = option::is_some_and!(None, || false);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::is_some_and` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:73:44
   |
73 | const _: bool = option::is_some_and!(None, {whatever1_return});
   |                                            ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |                 ^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:75:17
   |
75 | const _: bool = option::is_none_or!();
   |                 ^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$opt:expr`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:76:42
   |
76 | const _: bool = option::is_none_or!(None,);
   |                                          ^ missing tokens in macro arguments
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |                 ^

error: expected the closure to take a pattern as an argument
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:77:17
   |
77 | const _: bool = option::is_none_or!(None, || false);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `option::is_none_or` (in Nightly builds, run with -Z macro-backtrace for more info)

error: no rules expected `{`
  --> tests/misc_tests/misc_macro_ui_tests/option_macro_syntax_err.rs:78:43
   |
78 | const _: bool = option::is_none_or!(None, {whatever1_return});
   |                                           ^ no rules expected this token in macro call
   |
note: while trying to match `|`
  --> src/option.rs
   |
   |     ($opt:expr, |$param:pat_param| $pred:expr $(,)?) => {
   |                 ^
